pkg/proxy/types.go

```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
 func (spn ServicePortName) String() string {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+// ServicePort is an interface which abstracts a service information.
+type ServicePort interface {
+	// String returns service string.
+	String() string
+	// ClusterIP returns service cluster IP.
+	ClusterIP() string
+	// Protocol returns service protocol.
+	Protocol() api.Protocol
+	// HealthCheckNodePort returns service health check node port if present.  If return 0, it means not present.
+	HealthCheckNodePort() int
+}
+
+// Endpoints is an interface which abstracts an endpoints info.
+type Endpoints interface {
```

Comment:
What is the purpose of Endpoints interface and ServicePort interface? I'm a little confused on what you are trying to accomplish.

Reply:
See comments in https://github.com/kubernetes/kubernetes/issues/52464. Briefly speaking, interface `Endpoints` and `ServicePort` is used to abstract the current `serviceInfo` and `endpointsInfo`. If we keep them private variable, we are not able to do refactor.

Reply:
After the refactor, IPVS kube-proxy will share the same public Endpoints, ServicePort interfaces and their public methods.

Reply:
Thanks for clarifying :)